rfd/0159-webasset-cache.md

```diff
@@ -0,0 +1,164 @@
+
+---
+authors: Michael Myers (michael.myers@goteleport.com)
+state: draft
+---
+
+# RFD 0159 - Webasset Cache
+
+## Required approvers
+
+
+## What
+
+We discuss the creation of an in-memory cache of webassets in the auth server to
+provide a way for proxies to serve files from different versions of Teleport.
+Here is a quick PoC https://www.loom.com/share/9983cae0e8574149a92dcec623e07f10
+
+## Why
+
+Up until now we've always thought of the 1-to-1 relationship of webasset bundle
+to proxy version was a given. In most cases, it still is. However, some
+customers may run multiple versions of the proxy behind a load balancer and this
+would cause the web browser to sometimes not be able to load the correct
+webasset.
+
+The way the frontend is bundle is by using code splitting. In a non-code split
+app, the entire react bundle would be compiled into a single javascript file and
+served with the index.html. This leads to some pretty large page loads so the
+solution is code splitting, which automatically splits the javascript bundle
+into bite-sized (byte-sized 🤭) chunks that can be served only when needed (page
+load, component load, etc etc). These chunks are generally split by feature and
+the files are appended with hashes that change every build (every version). If a
+web client receives an index.html that will request a file such as
+`index-123123.js` and then later on during the session requests
+`index-234234.js`, if the proxy that is hit when load balancing is of a
+different version than the original that served the `index.html`, that proxy
+will not have the file and return a `404`. Storing all the webassets in a cache
+on auth will allow the proxies to fallback to auth as a single source of truth
+and serve any file needed.
+
+## Details
+
+### Webasset cache
+
+The webasset cache is a simple `map[string][]byte` that lives in the auth server.
+
+
+```go
+type WebassetCache struct {
+	webassets map[string][]byte
+}
+```
+
+This cache gets populated when a proxy comes online. The proxy will initialize
+and register and then emit a message for each one of it's webasset files. This
+includes everything in the embedded filesystem such as javascript, css, fonts,
+images, etc. The total size, uncompressed, of the bundles are around ~11mb for
+enterprise and slightly less for oss, but for the sake of discussion we can
+average it at around 10mb. This means the cache should be roughly 10mb per
+version stored. As of now, we only know of customers running two versions
+simultaneously.
+
+A simple example of the new grpc service would look like this
+
+```protobuf
+message GetWebassetRequest {
+  string name = 1;
+}
+
+message GetWebassetResponse {
+  string name = 1;
+  bytes content = 2;
+}
+
+service WebassetCacheService {
+  // GetWebasset
+  rpc GetWebasset(GetWebassetRequest) returns (GetWebassetResponse);
```

Comment:
Worth mentioning the maximum size of a gRPC response message here?

Reply:
yes. 4mb max message size. I will add that to the RFD